[dnf5] System state: store dependency state, from_repo, state for groups and modules

lukash opened the pull request:
It's a bit rough around the edges and the API is not entirely consistent. I'm still not sure about how the API should look, hence I'm trying out various approaches, I'd like to merge and then do a cleanup round later, when we see what works better.

Commit:
libdnf/system/state: Store dependencies, store from_repo for nevras

Changes the on-disk format of the system state. Stores the
`userinstalled` flag for all NAs, allowing to distinguish userinstalled
packages and dependencies.

Also stores `from_repo` (id of repository from which the package was
installed) for all NEVRAs.

Properly adds and removes RPM packages' state in
libdnf/base/transaction.

Removes a couple of installonly packages hacks, which should now be
properly handled by storing the state for NA and by checking for
presence of the NA on the system before removing the package state NA
record.

Commit:
libdnf/system/state: Store module state

Commit:
libdnf/system/state: Store group state

lukash commented:
The CI is failing on F35, because on F36+ toml11 outputs a leading empty line when serializing data. This is technically a regression, I should look into upstream to fix these minor nuisances, but at this stage I'd remove F35 from the CI, we're not targeting it with the release anyway.

j-mracek commented on libdnf/base/transaction.cpp:
```diff
@@ -326,10 +271,15 @@ TransactionPackage Transaction::Impl::make_transaction_package(
+
+    transaction::TransactionItemReason reason;
+    if (action == TransactionPackage::Action::INSTALL) {
+        reason = solved_goal.get_reason(id);
```
I checked the code and I believe that it will not work correctly for installonly packages. I guess that the proposed implementation will fail for tests: https://github.com/rpm-software-management/ci-dnf-stack/blob/master/dnf-behave-tests/dnf/installonly.feature scenario  `Scenario: Install multiple versions of an installonly package and keep reason`.

j-mracek replied:
This code is still required to cover some scenarios with installonly packages
```
        //  Inherit the reason if package is installonly an package with the same name is installed
        //  Use the same logic like upgrade
        //  Upgrade of installonly packages result in install or install and remove step
        if (installonly_query.p_impl->contains(id)) {
            rpm::PackageQuery query(installed_installonly_query);
            query.filter_name({pool.get_name(id)});
            if (!query.empty()) {
                reason = tspkg.get_package().get_reason();
            }
        }
```

lukash replied:
I've solved the problem with a bit different query (well, similarly to the inner query here, but filtering for NA instead of just name. Not sure how much of a performance issue this could be, but OTOH it's independent on whether a package is installonly. Though, maybe we should have a `Package::is_installonly()` method which would be helpful here?

j-mracek commented on libdnf/base/transaction.cpp:
```diff
@@ -277,33 +246,9 @@ void Transaction::Impl::set_transaction(rpm::solv::GoalPrivate & solved_goal, Go
-
-        // https://bugzilla.redhat.com/show_bug.cgi?id=1921063
-        // To keep a reason of installonly pkgs in DB for remove step it requires TSI with reason change
-        for (auto id : list_removes) {
```
I am really happy to see the removal of this code, because it is not the correct place where it should be handled. But I don't see the logic that replaces it. I know that we already discussed possible solution but I feel not quite comfortable because you corrected the code but it creates an issue in functionality.

j-mracek replied:
Or did I overlooked the new logic that replaced the functionality?

j-mracek replied:
I think I found the new logic in transaction.cpp but I have to check again.

j-mracek replied:
It is there therefore I am closing this conversation as resolve. I am sorry for the noise.

j-mracek commented on include/libdnf/system/state.hpp:
```diff
@@ -30,6 +32,41 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 
+class PackageState {
+public:
+    bool userinstalled{false};
```
Do we really need to support a default constructor with a default value? Why did you create a default value `false`? I supposed that the functional default should be `true` if we need to support a default constructor.

j-mracek replied:
The comment is also valid for following classes.

lukash replied:
The default value is only there so that the member variable is never uninitialized. In this case the intention is to always initialize the value in construction, but I wasn't sure what the public interface for this should be, so I didn't bother with an explicit constructor yet and relied on direct member initialization (or whatever it's called, I'm not sure right now).

I can add it if you insist, but this will need to be revised from API standpoint anyway, so I thought I'd just leave it for now.

j-mracek replied:
Yes, I understand the default value. I would prefer to have no default constructor because to use a default value it would be also a problem. But it is not a blocker.

j-mracek commented on libdnf/base/transaction.cpp:
```diff
@@ -466,9 +416,44 @@ Transaction::TransactionRunResult Transaction::Impl::run(
+
+                rpm::PackageQuery query(installed_query);
+                query.filter_name({tspkg.get_package().get_name()});
+                query.filter_arch({tspkg.get_package().get_arch()});
```
I am guessing now. In case I have 2 installonly package and I am removing both of them I guess that the code will keep the reason.

j-mracek replied:
Anyway your approach is in principle better then the previous logic. But we have to resolve all corner cases.

lukash replied:
It works for this case, I've now enabled a scenario that tests it.

j-mracek replied:
Perfect, I realized one thing byt it is a performance optimization. We have a filter that can replace 
```
                query.filter_name({tspkg.get_package().get_name()});
                query.filter_arch({tspkg.get_package().get_arch()});
```
See 
```

    /// Filter packages by their `name` and `arch` based on names and arches of the packages in the `package_set`.
    ///
    /// @param package_set      PackageSet with Package objects the filter is matched against.
    /// @param cmp              A comparison (match) operator, defaults to `QueryCmp::EQ`.
    ///                         Supported values: `EQ`, `NEQ`.
    /// @since 5.0
    void filter_name_arch(const PackageSet & package_set, libdnf::sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
```

If you have only package as an argument we can create a variant for the package.

lukash replied:
So far there's no filter for a single package on `PackageQuery` interface. I wouldn't add one randomly, if we want to have filters on `Package`s, I'd add them systematically.

j-mracek commented on libdnf/base/transaction.cpp:
```diff
@@ -466,9 +416,44 @@ Transaction::TransactionRunResult Transaction::Impl::run(
+
+                // for a REINSTALL, the remove needs to happen first, hence the reverse iteration of the for loop
+                system_state.remove_package_nevra_state(tspkg.get_package().get_nevra());
+            } else {
```
I know that it is the remaining option but what about to test it rather then guess what the package reason we have?

lukash replied:
Done.

j-mracek reviewed (changes requested):
It looks fine. I think I discovered some issues but I believe that they can be resolved.

lukash commented:
I've tweaked some things and made this work for installonly packages (upgrades, etc.) and also for packages installed outside dnf and hence having reason `None`. I've added some more tests for these cases to the CI stack PR. If you can think of any other cases, we should add them to the tests as well.

j-mracek commented on include/libdnf/system/state.hpp:
```diff
@@ -30,6 +32,41 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+
+class NevraState {
+public:
+    std::string from_repo;
```
For the NevraState state it quite easy to determine not initialized or a default value because the empty string is not valid repoid. Again it is not a blocker but somehow I feel it is not right.

lukash replied:
The class is not on the interface as of now, so the user can't make this mistake. It's just a dataholder. On the interface it's `void set_package_from_repo(const std::string & nevra, const std::string & from_repo);`. Not that it prevents passing an empty string... But it doesn't have the "uninitialized class" issue.

Again, this will need a bit of cleaning up later.

j-mracek commented on libdnf/transaction/transaction_item_reason.cpp:
```diff
@@ -65,15 +65,12 @@ TransactionItemReason transaction_item_reason_from_string(const std::string & re
-// TODO(lukash) the above doesn't make much sense. likely not needed with the new system state, revise
+
 static TransactionItemReason order[] = {
+    TransactionItemReason::NONE,
```
This part is used for goal to determine the proper reason with obsoletes. I think that the original position of none is correct.
```
        for (int i = 0; i < obs.size(); ++i) {
            rpm::Package obsoleted(base, rpm::PackageId(obs[i]));
            auto obs_reson = obsoleted.get_reason();
            if (obs_reson > reason) {
                reason = obs_reson;
            }
```

lukash replied:
It's not correct, it's a horrible hack, unfortunately. If you know of a case in which this will fail, please describe it so that I can create a test scenario for it.

As for a solution, I consider special-casing `None` in a particular case better than changing the otherwise logical order of the reasons. Note that now the `get_reason()` method returns `None` when there's no record for the NA in `packages.toml`. That can mean two things though, either the package is not installed, or it was installed outside dnf and the "effective" reason is `User`. I've even been considering adding a `get_effective_reason()` method to `Package`, but using that in the `base::Transaction` logic wouldn't work well AFAICS as that still erases the above-described distinction.

j-mracek replied:
Well thank you for your description because I realized that the issue is a different. The issue is that None reason represents two reasons and we have to distinguish between them. Please can you add a reason representing installed outside of DNF stack?

j-mracek replied:
The main user of reason comparison is Transaction class where only usage of None is for packages installed outside of DNF because the comparison only take place where installed package is replaced.

j-mracek commented on libdnf/base/transaction.cpp:
```diff
@@ -323,13 +273,28 @@ TransactionPackage Transaction::Impl::make_transaction_package(
+
+    transaction::TransactionItemReason reason;
+    if (action == TransactionPackage::Action::INSTALL) {
+        reason = std::max(new_package.get_reason(), solved_goal.get_reason(id));
```
This is good and I like it. I believe that `new_package.get_reason()` should return a different reason for not installed packages and for installed packaged installed outside of our stack. Also that would be the best place to put the logic because it is outside of system state but still internal. What do you think?

lukash replied:
I see what you mean, but I don't like it. The `TransactionItemReason` is already a hodge-podge of values that are only applicable in specific cases and this adds one more that is again special. In addition on a `Package` class level the case (installed outside dnf) is represented by `is_installed()` returning `true` and `get_reason()` returning `None`, so it's not that we don't have this easily covered on the API.

Again, if you know of cases for which the current implementation will fail, let me know so I can add them to the CI stack tests.

j-mracek replied:
The first issue with the code is already mentioned bellow - evaluation redundancy, but this is not a problem. I believe that the package `get_reason()` should distinguish between two different reasons that are now represented by the None reason. Not installed or installed outside of libdnf5 stack and also removed outside of libdnf5 stack (present entry in system state but package is not installed anymore). It has also one advantage we do not need to test None reason in code, like it is done on following lines. The lines bellow did not make a decision how None will be handled therefore somewhere else the decision will be made and there will be a suggestion that None means modified outside of DNF. The huge advantage will be for repertory command where installed outside of DNF5 will be explicitly visible as a specific reason.

j-mracek replied:
I also think that the code will have a problem with handling of obsoletes on line 303 `if (old_reson > reason) {`.

I took a scenario https://github.com/rpm-software-management/ci-dnf-stack/blob/master/dnf-behave-tests/dnf/obsoletes.feature and I modified install line to install it with rpm.
```
@dnf5
# TODO(nsella) transaction table output
Scenario: Autoremoval of obsoleted package
   When I execute rpm with args "install PackageB-1.0"
   Then the exit code is 0
   When I execute dnf with args "upgrade"
   Then the exit code is 0
    And Transaction is following
        | Action        | Package                                   |
        | install       | PackageB-Obsoleter-0:1.0-1.x86_64         |
        | obsoleted     | PackageB-0:1.0-1.x86_64                   |
   When I execute dnf with args "autoremove"
   Then the exit code is 0
    But Transaction is empty
```
I guess that autoremove will remove obsoleted package as unused dep. It is because libsolv for such a tranaction will return `dep` and `PackageB` installed reason is None and None is lover then Dep. I gues that similar problem is also with upgrade depependency that obsoletes package with None reason.

j-mracek replied:
We can also discuss what is expected the return value for available packages for `new_package.get_reason()` - whether it is None or a reason of installed packages with the same NA.

j-mracek replied:
There is one additional eddge case with installroot packages. Imagine situation when I will install one installonly package (version 1) (requested by user therefore userinstalled) and the second one will be injected into transaction as a dependency (version 2) or obsoleter or obsoleter package installed outside of DNF5. Then during processing of packages in installset than the order of them will result effect the final reason. It can be resolved by additional sorting of transaction items or during updating System state, but both code requires reason comparison.

lukash replied:
> I also think that the code will have a problem with handling of obsoletes on line 303 `if (old_reson > reason) {`.
> 
> I took a scenario https://github.com/rpm-software-management/ci-dnf-stack/blob/master/dnf-behave-tests/dnf/obsoletes.feature and I modified install line to install it with rpm.
> 
> ```
> @dnf5
> # TODO(nsella) transaction table output
> Scenario: Autoremoval of obsoleted package
>    When I execute rpm with args "install PackageB-1.0"
>    Then the exit code is 0
>    When I execute dnf with args "upgrade"
>    Then the exit code is 0
>     And Transaction is following
>         | Action        | Package                                   |
>         | install       | PackageB-Obsoleter-0:1.0-1.x86_64         |
>         | obsoleted     | PackageB-0:1.0-1.x86_64                   |
>    When I execute dnf with args "autoremove"
>    Then the exit code is 0
>     But Transaction is empty
> ```
> 
> I guess that autoremove will remove obsoleted package as unused dep. It is because libsolv for such a tranaction will return `dep` and `PackageB` installed reason is None and None is lover then Dep. I gues that similar problem is also with upgrade depependency that obsoletes package with None reason.

Okay, this case is essentially the same as the one I came up with. The reason for this one actually ends up being "User", i.e. the following scenario passes (note the `package state is` step at the end):
```gherkin
@dnf5
Scenario: Autoremoval of obsoleted package
   When I execute rpm with args "-i --nodeps {context.scenario.repos_location}/dnf-ci-obsoletes/x86_64/PackageB-1.0-1.x86_64.rpm"
   Then the exit code is 0
   When I execute dnf with args "upgrade"
   Then the exit code is 0
    And Transaction is following
        | Action        | Package                                   |
        | install       | PackageB-Obsoleter-0:1.0-1.x86_64         |
        | obsoleted     | PackageB-0:1.0-1.x86_64                   |
   When I execute dnf with args "autoremove"
   Then the exit code is 0
    But Transaction is empty
    And package state is
        | package                         | userinstalled | from_repo        |
        | PackageB-Obsoleter-1.0-1.x86_64 | True          | dnf-ci-obsoletes |
```

This is because the action for the package is `INSTALL` and the reason is taken from the goal, which gives `USER` as reason.

I'm thinking of a potential solution to unconditionally force `NONE` as the final reason in case one of the obsoleted packages has reason `NONE`...

j-mracek commented on libdnf/base/transaction.cpp:
```diff
@@ -323,13 +273,28 @@ TransactionPackage Transaction::Impl::make_transaction_package(
+        // For installonly packages: if the NA is already on the system, but
+        // not recorded in system state as installed, it was installed outside
+        // DNF and we want to preserve NONE as reason
+        if (!installed_query.empty() && new_package.get_reason() == transaction::TransactionItemReason::NONE) {
```
If we will move the logic to `new_package.get_reason()` we can remove above lines. What do you think?

j-mracek replied:
Anyway this line is somehow redundant because it reverts decision from line 284 using the same set of information plus result from the mandatory `installed_query`.

pkratoch commented on libdnf/module/module_item_container.cpp:
```diff
@@ -70,4 +71,37 @@ BaseWeakPtr ModuleItemContainer::get_base() const {
+    : libdnf::Error(M_("Invalid module state: {}"), state) {}
+
+
+std::string module_state_to_string(ModuleState state) {
+    switch (state) {
+        case ModuleState::AVAILABLE:
+            return "Available";
+        case ModuleState::DEFAULT:
+            return "Default";
+        case ModuleState::ENABLED:
+            return "Enabled";
+        case ModuleState::DISABLED:
+            return "Disabled";
+    }
+    return "";
+}
+
+
+ModuleState module_state_from_string(const std::string & state) {
+    if (state == "Available") {
+        return ModuleState::AVAILABLE;
+    } else if (state == "Default") {
+        return ModuleState::DEFAULT;
+    } else if (state == "Enabled") {
+        return ModuleState::ENABLED;
+    } else if (state == "Disabled") {
+        return ModuleState::DISABLED;
+    }
+
+    throw InvalidModuleState(state);
+}
```
In DNF 4, there was empty string for both `AVAILABLE` and `DEFAULT` states, because it doesn't make sense to distinguish between these two states in the config (it depends on the loaded metadata and can change over time). And the strings that are used e.g. in `module list` or `module info` when printing states of modules are different anyway, so these `module_state_to_string` and `module_state_from_string` are only used for the `system::ModuleState`. I think it would be better to have one string for both of these states and interpret it as `AVAILABLE`.

j-mracek replied:
I guess that the information is even stored internally on different places.

lukash replied:
@pkratoch well that's not very good... in fact, that's pretty bad :slightly_smiling_face: Adding this sort of inconsistency is a really bad idea. This rather tells me the enum should not contain one of the values, probably the `DEFAULT` one. I'm not sure how it's used, but would it possibly be better to store that attribute (a module having a default stream) separately, e.g. like `bool has_default_stream()`?

j-mracek commented on libdnf/base/transaction.cpp:
```diff
@@ -299,9 +299,12 @@ TransactionPackage Transaction::Impl::make_transaction_package(
+        auto old_reason = replaced_pkg.get_reason();
+        if (old_reason == transaction::TransactionItemReason::NONE &&
+            reason <= transaction::TransactionItemReason::DEPENDENCY) {
+            reason = transaction::TransactionItemReason::NONE;
```
I am really sorry but I prefer other solution that will modify comparison algorithm rather that to not use `std::max(reason, old_reason);` and replace it by other logic fore some conditions. Before I start with explanation why I recommend other solution I have to mentioned that the problem appeared after modification of comparison order in this PR.

I propose to resolve the reason value for comparison in lower level rather to creates every time the reason comparison is used to add additional code that will fix the comparison for some cases.

One option how to delivery required behavior would be to split NONE reason to two reasons - NONE and INSTALLED_BY_DIFFERENT_TOOL. NONE can be used for all available packages.

I think that the solution is 

- More transparent because the reason will explicitly says what it means (improvement to DNF4)
- System state will record INSTALLED_BY_DIFFERENT_TOOL as userinstalled=1 but history DB will store INSTALLED_BY_DIFFERENT_TOOL
- This will be less confusing for installonly packages because can be installed in multiple copies without removal
- Better understandable outside of code of goal (repoquery) and in history DB
- Performance - for available packages it will not require any calculation (looking into dictionary of system state, no RPM queries). It is remarkable because available packages represents 29/30 in our data set
- Bulletproof solution - anytime when anyone will try to compare reasons - it will not require any additional code around to cover some cases. We already discussed that such a comparison is required 3 times in code and we already talked about additional usage for sorting of transaction elements.
- Simplification - the code will be only represented by a single line - `reason = std::max(reason, old_reason);`

lukash replied:
The issue with your design is that it combines another boolean property, `Package::is_installed()` into a new value of the `Reason` enum.

Effectively, `pkg.get_reason() == INSTALLED_OUTSIDE_DNF` is equivalent to `pkg.is_installed && pkg.get_reason() == NONE`. I am convinced this alone is an obvious proof of a bad design. In general, if there's a need for such combination for convenience (or performance) reasons, the two elementary values should be left untouched and a third, compound value should be introduced that combines them into another semantically important attribute.

The `NONE` value as of now represents that there is no record of the user selecting a reason for a package in the system state toml. You are effectively splitting the value into two, with your new reason, if someone on the API would want to find out if a package has no reason, they will need the following condition: `pkg.get_reason() == NONE || pkg.get_reason() == INSTALLED_OUTSIDE_DNF`.

As for the argument of it being faster with not having to query the system repo every time you need to distinguish the two, instead there's the added complexity of the system state class needing to have the system repo as an input so that it can query, precompute and cache the information. Since it's effectively being a cache for this, there are also other ways to cache this without needing to add another enum value on the interface.

> ...
> I think that the solution is
>
>     * More transparent because the reason will explicitly says what it means (improvement to DNF4)
>
>     * System state will record INSTALLED_BY_DIFFERENT_TOOL as userinstalled=1 ...

No it won't, it'd still be missing a record for that NA, otherwise we'd erase the distinction and the reason would become USERINSTALLED. This is fine though and not an issue with the approach (AFAICS).

>     * ... but history DB will store INSTALLED_BY_DIFFERENT_TOOL

Normal history DB records won't have this. Only the "differential transactions" that we talked about (automatically created records of unexpected system state diffs we detect) could have it, but we could just as well use NONE for it.

>     * This will be less confusing for installonly packages because can be installed in multiple copies without removal

I don't understand this point.

>     * Better understandable outside of code of goal (repoquery) and in history DB

I don't think it makes a difference in history DB. NONE can represent the information just as well.

As for better understandable code, it depends on what information you need to retrieve. If you need to know if there's no reason record for a package, you'll now need to check for two enum values instead of one, which is worse.

>     * Performance - for available packages it will not require any calculation (looking into dictionary of system state, no RPM queries). It is remarkable because available packages represents 29/30 in our data set

I've addressed this above with the "cache" argument.

>     * Bulletproof solution - anytime when anyone will try to compare reasons - it will not require any additional code around to cover some cases. We already discussed that such a comparison is required 3 times in code and we already talked about additional usage for sorting of transaction elements.

I'd prefer you didn't use the word bulletproof in the context of dnf reasons... it's a house of cards one way or the other.

The need for the workarounds is because dnf has the convention of treating INSTALLED_OUTSIDE_DNF as userinstalled. That's the real cause of this issue.

I'm not even sure the comparator should be on the API, IIRC when we discussed it you said it's supposed to only be internal.

>     * Simplification - the code will be only represented by a single line - `reason = std::max(reason, old_reason);`

This simplification here comes at the cost of more complexity elsewhere.

It's not that I don't see the advantages of having the extra value, but I consider the benefits marginal, the complexity comparable (just in different places) and the design issues I outlined simply push the needle into the red area for me.

j-mracek replied:
It would be nice to see the issue with reasons in a bigger perspective. Or from history db view. When I see record in history DB with None reason I will suggest that it is bug in DNF. Also I can consider when I install packages as weak dependency than in the next upgrade transaction the  reason will change to dependency. Also the interesting part is when package will be installed as group and as a user but history structure does not support that complexity therefore we have another difference. And there is the last aspect, we have not much time left for development. It means that we have huge inconsistency between design of history DB and system state.

j-mracek commented:
I checked our CI and I think that related scenario to the PR are
```
install-obsoletes.feature
installonly.feature
installonlypkgs.feature
obsoletes.feature
autoremove.feature
```
Additionally there are scenarios related to groups and modules.

Commit:
libdnf/utils/sqlite3: Copy and free expanded_sql immediately

The `get_expanded_sql()` method returning a char * that is freed in the
Statement destructor was causing an error in exception classes, which
store the char * and use it after the stack is unwound and the Statement
is destroyed. We could copy the string into the errors but if a copy has
to be made, it's better to just return a std::string and free the char *
right away.

j-mracek commented on libdnf/rpm/package.cpp:
```diff
@@ -313,7 +314,23 @@ std::string Package::get_repo_id() const {
 
 libdnf::transaction::TransactionItemReason Package::get_reason() const {
-    return base->get_system_state().get_package_reason(get_na());
+    auto reason = base->get_system_state().get_package_reason(get_na());
```
I prefer to add a test whether package with NA is installed first. I think it will provide a better behavior.

lukash replied:
I am not sure how you mean it exactly, did you mean it like this?
```c++
rpm::PackageQuery installed_query(base, rpm::PackageQuery::ExcludeFlags::IGNORE_EXCLUDES);
installed_query.filter_installed();
installed_query.filter_name({get_name()});
installed_query.filter_arch({get_arch()});
if (!installed_query.empty()) {
    auto reason = base->get_system_state().get_package_reason(get_na());

    if (reason == libdnf::transaction::TransactionItemReason::NONE) {
        return libdnf::transaction::TransactionItemReason::EXTERNAL_USER;
    }

    return reason;
}

return libdnf::transaction::TransactionItemReason::NONE;
```

I don't agree it's better, for one, it is quite a bit more expensive to always do the query, instead of only when the reason is `NONE` (a small minority of cases). Of course, in this case you don't have to search for the reason if the NA is not installed, but the assumption here is that most of the time you're querying for a NA that is installed.

Second, it changes behavior, I've created a test case for this: https://github.com/rpm-software-management/ci-dnf-stack/pull/1128/commits/a3a115cbbac414847f7ee3d04d823f679bf341da

Using the above block will break the test case, the `filesystem.x86_64` will have reason `Dependency` after the second install.

j-mracek replied:
> Using the above block will break the test case, the filesystem.x86_64 will have reason Dependency after the second install.

Do you know why it returns dep?

lukash replied:
> Do you know why it returns dep?

Yes, because with your approach it checks whether the NA is installed first, and it's not. There's reason `USER` in system state, but we check for the NA being installed and since it's not, we return `NONE`.

j-mracek replied:
I am sorry I overlooked something in the test. DEP is correct for `filesystem.x86_64` because the test requests to install only `basesystem` and `filesystem` was added to transaction as a dependency.

lukash replied:
Ok. Do you then want me to use the piece of code mentioned in https://github.com/rpm-software-management/libdnf/pull/1543#discussion_r933517720?

j-mracek replied:
Yes, if you can use it.

lukash commented:
Moved to the new repo, the PR is now here: https://github.com/rpm-software-management/dnf5/pull/1
